Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp

```diff
@@ -57,94 +58,67 @@ ProvisionalFrameProxy::ProvisionalFrameProxy(WebFrameProxy& frame, Ref<WebProces
 {
     m_process->markProcessAsRecentlyUsed();
     m_process->addProvisionalFrameProxy(*this);
-
-    m_process->addMessageReceiver(Messages::WebPageProxy::messageReceiverName(), m_pageID, *this);
-
+    // Its possible that this is called with a WebFrameProxy which has the message already registered (SiteIsolation.IframeNavigatesSelfWithoutChangingOrigin)
+    if (frame.processIdentifier() == m_process->processIdentifier()) {
+        m_process->removeMessageReceiver(Messages::WebFrameProxy::messageReceiverName(), m_frame.frameID().object());
+        m_process->removeFrameWithRemoteFrameProcess(frame);
+    }
     m_process->addMessageReceiver(Messages::WebFrameProxy::messageReceiverName(), m_frame.frameID().object(), *this);
 
     ASSERT(m_frame.page());
     auto& page = *m_frame.page();
     auto* drawingArea = page.drawingArea();
     ASSERT(drawingArea);
 
-    auto parameters = page.creationParameters(m_process, *drawingArea);
-    parameters.subframeProcessFrameTreeInitializationParameters = { {
-        frame.frameID(),
-        *page.frameTreeCreationParameters(),
-        m_layerHostingContextIdentifier
-    } };
-    parameters.isProcessSwap = true; // FIXME: This should be a parameter to creationParameters rather than doctoring up the parameters afterwards.
-    parameters.topContentInset = 0;
-    m_process->send(Messages::WebProcess::CreateWebPage(m_pageID, parameters), 0);
-    m_process->addVisitedLinkStoreUser(page.visitedLinkStore(), page.identifier());
-
-    drawingArea->attachToProvisionalFrameProcess(m_process);
+    if (newProcess) {
+        auto parameters = page.creationParameters(m_process, *drawingArea);
+        parameters.subframeProcessFrameTreeInitializationParameters = { {
+            frame.frameID(),
+            *page.frameTreeCreationParameters(),
+            m_layerHostingContextIdentifier
+        } };
+        parameters.isProcessSwap = true; // FIXME: This should be a parameter to creationParameters rather than doctoring up the parameters afterwards.
+        parameters.topContentInset = 0;
+        m_process->send(Messages::WebProcess::CreateWebPage(m_pageID, parameters), 0);
+        m_process->addVisitedLinkStoreUser(page.visitedLinkStore(), page.identifier());
+        drawingArea->attachToProvisionalFrameProcess(m_process);
+    }
 
     LoadParameters loadParameters;
```

Comment:
This needs to be moved to a separate function, not the constructor.

Reply:
based on offline convo, we can clean this up later if we think of a different design.